Add an example of text-image dual encoder model

ksalama opened the pull request:
Hello @fchollet and @8bitmp3 -

I created this example demonstrates how to build a dual encoder (or two-tower)
neural network model to search for images using natural language. The idea is to train a vision
encoder and a text encoder to project the representation of the images and their
captions into the same embedding space, such that the caption embeddings are near
to the embeddings of images describing them. I am using BERT from `TensorFlow Hub` as the base for the text encoder, and Xception from `keras.application` as the base for the vision encoder. I am also using the [MS-COCO](https://cocodataset.org/#home) image captioning dataset.

The implementation is relatively straightforward, but I kind of proposed a new technique to compute the loss, which I called  Contrastive Similarity Loss (let me know if you prefer another name).  The results are also interesting, but the training is a bit expensive :-)

Let me know what do you think. 

Cheers

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,605 @@
+"""
+
+
+class ContrastiveSimilarityLoss(keras.losses.Loss):
```
Given the fact that `y_true` is not used, I'm not sure a `Loss` subclass is the best option. A more natural factoring would be a custom `train_step` for the `dual_encoder` model. That said, the current factoring is still pretty simple, so I'll leave the decision up to you.

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,605 @@
+"""
+
+
+def create_dual_encoder(text_encoder, vision_encoder):
```
This could be a model subclass taking the 2 submodels in its constructor and featuring a custom `train_step`. The model has basically no topology so it would make a lot of sense. Its `call` pass would just be:

```
def call(self, captions, images):
    # Get the embeddings for the captions.
    caption_embeddings = self.text_encoder(captions)
    # Get the embeddings for the images.
    image_embeddings = self.vision_encoder(images)
    # Stack the caption_embeddings and image_embeddings to produce
    # the output
    return tf.stack([caption_embeddings, image_embeddings])
```

The `train_step` would be the content of the contrastive loss. You may have to implement an `evaluate_step` and `predict_step` as well.

fchollet replied:
I might recommend `def call(self, inputs)` then unpacking `captions`/`images` from the inputs, which enables `predict` to be used within a custom `predict_step`.

ksalama replied:
Thank you for the suggestion. I tried the custom model/training loop approach, but it doesn't be really needed, nor does it simplify the code. 

Note that the dual encoder is discarded once  the two encoders are trained, so we don't need the `predict_step` nor `evaluate_step`. The reasons is that, in inference time, we don't provide a query and an image to we predict their similarity (in which case, the `predict_step` would make sense), rather we provide a query and we search for the most similar image. So forward pass of the training process is irrelevant in prediction.

What do you think?

fchollet replied:
I think we should do this: https://gist.github.com/fchollet/196b4b48ce11569aa5e256dc6f274f6b (caveat: I haven't tested the code)

It is more extensible, and it does not involve any "hacks" (like fake labels), so it will look nicer to people who want maximum flexibility. It also takes fewer lines of code.

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,605 @@
+
+"""
+Note that training the model with 60,000 image-caption pairs takes around 15 minutes
+per epoch using a P100 GPU accelerator"""
```
This is very dependent on the batch size, and potentially on the `steps_per_execution` parameter (how many batches to process in a single graph execution)

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,605 @@
+
+print(f"Image embeddings generation started...")
+print(f"Generating embeddings for {len(image_paths)} images...")
+while idx < len(image_paths):
```
You can dramatically simplify this code by calling `vision_encoder.predict(dataset)` and putting your images in a dataset (could be done via `image_dataset_from_directory`)

You could also simply create a dataset from your image paths, then map `read_file`, `decode_jpeg`, `resize`, to the dataset, and batch it. Maybe than would be even simpler than `image_dataset_from_directory`.

ksalama replied:
Done. Is what what you had in mind?

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,605 @@
+between the input query embedding and the image embeddings, and retrieve the top k
+matches. However, *approximate* similarity matching, using frameworks like
+[ScaNN](https://github.com/google-research/google-research/tree/master/scann),
+[Annoy](https://github.com/spotify/annoy), or [Faiss](https://github.com/facebookresearch/faiss)
```
Fun fact: "faiss" would be pronounced like "fesse" (ass cheek) in French.

ksalama replied:
Hahaha, I know!!! (I speak french a little bit as my wife is French).

fchollet reviewed (commented):
This is a very interesting model -- great stuff! 👍

ksalama commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,605 @@
+import matplotlib.pyplot as plt
+import matplotlib.image as mpimg
+
+tf.get_logger().setLevel("ERROR")
```
The tf.hub module produces a lot of warnings, therefore I added this. 

Its seems that tf.hub set the logging level to error in their [examples](https://www.tensorflow.org/tutorials/text/classify_text_with_bert)

What do you think?

ksalama replied:
Done

ksalama commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,605 @@
+Title: Natural Language Search with Dual Encoder
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/30
+Last modified: 2021/01/30
```
Ok. Will update when I generate the other files

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,612 @@
+print(f"Batch size: {batch_size}")
+print(f"Steps per epoch: {train_example_count // batch_size}")
+train_data = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+history = dual_encoder.fit(train_data, epochs=num_epochs)
```
It would be useful to use validation data to monitor overfitting, no?

ksalama replied:
Good point. Added.

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,612 @@
+
+print(f"Image embeddings generation started...")
+print(f"Generating embeddings for {len(image_paths)} images...")
+for image_batch in (
```
This paragraph can be replaced with

`image_embeddings  = vision_encoder.predict(dataset)`

no?

ksalama replied:
Nice one! Thanks

fchollet reviewed (commented):
Thanks for the update! How many lines of code does the example have at this time?

ksalama commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,609 @@
+    image = tf.image.decode_jpeg(raw_image, channels=3)
+    image = tf.image.resize(image, (299, 299))
+    features["image"] = image
+    return features, tf.constant(1)
```
The `model.fit` expects a tuple of (inputs, targets). Since the targets are dynamically computed during loss calculation, a tf constant is added as a placeholder for the targets. I added this a comment for clarification.

ksalama commented:
@fchollet - Thanks for your suggestions. I tried to use the `autogen` tool and it didn't complain about the number of lines. So I think we are good. I will test your approach (thanks for sharing the gist) and update the PR. Cheers.

ksalama commented:
@fchollet - I have updated the implementation with your approach. It works fine (although I kind of prefer avoiding implementing custom training loop if possible, and just implement the required logic for the model, which is the loss in our case). I also added early stopping and lr_reducer.

Let me know if you have any further comments

ksalama commented:
Now added support for model parallelism by placing each encoder to a separate GPU. This gained about 33% time reduction.

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+"""
+Title: Natural Language Search with Dual Encoder
```
Suggestion:

```suggestion
Title: Natural Language Search with a Dual Encoder
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/30
+Last modified: 2021/01/30
+Description: Implementing a dual encoder model for image search with natural language.
```
Suggestion:

```suggestion
Description: Implementation of a dual encoder model for image search with natural language.
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+## Prepare the data
+
+We will use the [MS-COCO](https://cocodataset.org/#home) dataset to train our
+dual encoder model. It contains over 82,000 images, each of which has at least
```
Suggestion:

"It" may appear to be referring to the "model", because of the order of the words in the previous sentence, so:

```suggestion
dual encoder model. MS-COCO contains over 82,000 images, each of which has at least
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+###
+Download and extract the data
+
+First, let's download the dataset, which consists of two compressed folders.
+One contains the images, and the other contains their captions.
```
Suggestion:

Repetition

```suggestion
First, let's download the dataset, which consists of two compressed folders:
one with images, and the other—with associated image captions.
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+
+First, let's download the dataset, which consists of two compressed folders.
+One contains the images, and the other contains their captions.
+Note that the compressed images folder is 13GB.
```
```suggestion
Note that the compressed images folder is 13GB in size.
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+
+
+"""
+## Implement projection head
```
Suggestion:

```suggestion
## Implement the projection head
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+"""
+## Implement the text encoder
+
+In this example, we use [BERT](https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-256_A-4/1)
```
Suggestion:

Repetition

```suggestion
We use [BERT](https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-256_A-4/1)
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+The target similarity between `caption_i`  and `image_j` is computed as
+the average of the (dot-product similarity between `caption_i` and `caption_j`)
+and (the dot-product similarity between `image_i` and `image_j`).
+Then we use crossentropy to compute the loss between the targets and the predictions.
```
Suggestion:

```suggestion
Then, we use crossentropy to compute the loss between the targets and the predictions.
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+"""
+## Train the dual encoder model
+
+In this experiment, we freeze the base encoders for text and image, and make only
```
Suggestion:

```suggestion
In this experiment, we freeze the base encoders for text and images, and make only
```
(or "each image")

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+in the index to retrieve the indices of the top matches.
+4. Look up the paths of the top matching images to display them.
+
+Note that, after training the `dual encoder`, the only the fine-tuned `vision_encoder`
```
Suggestion:

```suggestion
Note that, after training the `dual encoder`, only the fine-tuned `vision_encoder`
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+### Generate embeddings for the images
+
+We load the images and feed them into the `vision_encoder` to generate their embeddings.
+In large scale systems, this step is performed using a parallel data processing frameworks,
```
Suggestion:

```suggestion
In large scale systems, this step is performed using a parallel data processing framework,
```

8bitmp3 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+We load the images and feed them into the `vision_encoder` to generate their embeddings.
+In large scale systems, this step is performed using a parallel data processing frameworks,
+such as [Apache Spark](https://spark.apache.org) or [Apache Beam](https://beam.apache.org).
+Generating the image embeddings will take a few minutes.
```
Suggestion:

```suggestion
Generating the image embeddings may take several minutes.
```

8bitmp3 reviewed (commented):
👍 PTAL

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+print(f"Steps per epoch: {int(np.ceil(train_example_count / batch_size))}")
+train_dataset = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+valid_dataset = get_dataset(os.path.join(tfrecords_dir, "valid-*.tfrecord"), batch_size)
+# Create a learning rate scheduler callback.
```
Have you tried one of the built-in learning rate schedules? In general I'd recommend using one of these rather than a callback. https://keras.io/api/optimizers/learning_rate_schedules/

ksalama replied:
I came across it as I found that such technique (reducing learning rate if eval quality doesn't improve) is being used in several recent research papers. I am not sure how effective this technique in this particular example, but I think it is cool and I wanted to highlight it :-)

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+    validation_data=valid_dataset,
+    callbacks=[reduce_lr, early_stopping],
+)
+print("Training completed. Saving vision and text encoders...")
```
Q: what happens if you save the joint model -- does it work as expected?

ksalama replied:
Yes it does. I Just wanted to make the point that the joint model (Dual encoder) will not be used for serving. We need the `vision_encoder` to generate embeddings for your image library (one-off), and the text_encoder to generate embedding for the input queries (online serving).

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+    return tf.image.resize(image_array, (299, 299))
+
+
+print(f"Image embeddings generation started...")
```
Nit: no need for the `f"` here

ksalama replied:
Done

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+    tf.data.Dataset.from_tensor_slices(image_paths).map(read_image).batch(512),
+    verbose=1,
+)
+print(f"Image embeddings generation completed.")
```
Same here (and in a few more places in this file)

ksalama replied:
Done

fchollet commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,635 @@
+
+    def call(self, features, training=False):
+        # Place each encoder on a separate GPU (if available).
+        with tf.device("/gpu:0"):
```
The device scopes are hardcoding the assumption that the runtime has 2 GPUs, which is not a safe assumption. We should make this optional.

Besides I'm not sure putting the device scopes in `call` is the right approach wrt variable placement. Consider instead creating the models (e.g. calling `create_vision_encoder`) under the device scope.

ksalama replied:
I tried the code with a machine with no GPUs, 1 GPU, and 2 GPUs, and in all cases the code worked fine. It seems like tf falls back to the available devices if it cannot find the device in the scope. So I think it is safe to leave it as-is.

Besides, I tried to define the scope in the creation:
```
with tf.device("/gpu:0"):
    vision_encoder = create_vision_encoder(
        num_projection_layers, projection_dims, dropout_rate
    )
```
But it didn't work as expected (with 2 GPUs, only one was utilized). Do you mean to define the scope somewhere else?

In fact, in the [Keras FAQ](https://keras.io/getting_started/faq/) regarding model parallelism, the example shows the scope definition around calling the layer/model (btw, the example should be updated to change `tf.device_scope` to `tf.device`)

Let me know your thoughts.

fchollet replied:
Ok, if it works, then sounds good. Please add a comment explaining that TF will fallback on available devices if there are fewer than 2 GPUs.

fchollet reviewed (commented):
Looking great!

@8bitmp3 thank you for the helpful suggestions.

ksalama commented:
Thanks @8bitmp3 and @fchollet for your reviews. I incorporated your suggestion.

fchollet reviewed (approved):
LGTM, thank you!

fchollet commented:
You can add the generated files (after adding the comment about devices scopes)

ksalama commented:
I am working on it (had to reduce the number of code lines due to the subclassing approach, and also reduce the number of epochs so that the generation of the files doesn't timeout).👍

ksalama commented:
Done - The files are generated.
Although the model was trained for only 5 epochs, the matching images for the sample query are not bad! (however, the evaluation results are not great, of course)

fchollet commented:
Thank you for the great contribution! 👍 I'll take it from here.

lc0 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,588 @@
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_hub as hub
+import tensorflow_text as text
```
`tensorflow_text` is not really used, since preprocessing from BERT is used directly. Or am I missing something?

fchollet replied:
As a side-note, when importing `tensorflow_text` I'd recommend naming it `tf_text`, since `text` can be pretty confusing

lc0 commented on examples/nlp/nl_image_search.py:
```diff
@@ -0,0 +1,588 @@
+    monitor="val_loss", factor=0.2, patience=3
+)
+# Create an early stopping callback.
+early_stopping = tf.keras.callbacks.EarlyStopping(
```
keras is already imported from tensorflow. For sake of consistency with other callback just a few lines above we can drop `tf.` part
